### PR TITLE
the new hotness

### PR DESCRIPTION
**Pull request by apiarian**

### CODE REVIEW

File: shiny-api/updater/updater.go [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+	"sync"
+	"time"
+
+	"github.com/apiarian/migration-playground/shiny-api/common"
+)
+
+type Updater struct {
+	kc            *common.KafkaClient
+	command_topic string
+	done          chan struct{}
```
Comment (kklipsch) [inline shiny-api/updater/updater.go]: done channels have been surpassed in go with the context.Context object.

File: shiny-api/updater/updater.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	return u.kc.RegisterMessageProcessor(
+		u.command_topic,
+		5*time.Minute,
+		u.done,
+		common.WrapThingCommandEntryProcessor(u.MakeCommandProcessor()),
+	)
+}
+
+func (u *Updater) MakeCommandProcessor() func(*common.ThingCommandEntry, error) {
+	return func(m *common.ThingCommandEntry, err error) {
```
Comment (kklipsch) [inline shiny-api/updater/updater.go]: A common pattern is to have this run on a single non-dying goroutine.  Then you communicate with it via channels.  That way the internals to this function needn't be syncronized (and given you have fairly detailed sync logic, that seems a significant advantage).

File: shiny-api/updater/updater.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+		if err != nil {
+			log.Printf("error before command could be processed: %v", err)
```
Comment (kklipsch) [inline shiny-api/updater/updater.go]: Generally speaking, we've found that logging on a hotpath is an anti pattern in evented systems as you can get *lots* of events and see logs fill up.

File: shiny-api/updater/updater.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+					CreatedOn: m.Request.CreatedOn,
+					UpdatedOn: m.Request.UpdatedOn,
+					Version:   m.Request.Version,
+				}
+
+				u.mux.Lock()
+				defer u.mux.Unlock()
+				_, exists := u.entityCache[id]
+				if exists {
+					err := u.kc.PublishThingErrorCommand(
```
Comment (kklipsch) [inline shiny-api/updater/updater.go]: So here you have one of the first classic choices.  Do you retry the publish or fail it.  If you fail it you have data loss.  if you retry it, you have the chance of multiple publishes.  Generally speaking we prefer retries with data that is idempotent.

File: shiny-api/updater/updater.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+		}
+
+		if m.Request != nil {
+			switch m.Type {
+			case common.CreateCommandType:
+				var id string
+				if m.Request.ID == "" {
+					if u.assignIDs {
+						u.mux.Lock()
+						i := u.nextID
```
Comment (kklipsch) [inline shiny-api/updater/updater.go]: This makes your server essentially unscalable as its using an in memory id that only this process can know about.

File: shiny-api/updater/updater.go [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+	nextID int,
+) *Updater {
+	return &Updater{
+		kc:            kc,
+		command_topic: command_topic,
+		done:          make(chan struct{}),
+		assignIDs:     assignIDs,
+		nextID:        nextID,
+		mux:           &sync.Mutex{},
+		entityCache:   make(map[string]*common.Thing),
```
Comment (kklipsch) [inline shiny-api/updater/updater.go]: We've found it unlikely in the entities we deal with, that you should keep them in memory.  For this example its probably ok (and maybe if your service only deals with things down to the agencies level you'll never deal with a dataset that is big).

File: shiny-api/common/kafka.go [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+		"cid:'%s' type:%s request:%+v success:%+v failure:%+v",
+		te.CID,
+		te.Type,
+		te.Request,
+		te.Success,
+		te.Failure,
+	)
+}
+
+type KafkaClient struct {
```
Comment (kklipsch) [inline shiny-api/common/kafka.go]: Generally speaking, in golang methods are only preferred when you are messing with internal state as a side effect (or satisfying an interface).  Given that you aren't really doing much of that, it probably makes more sense to move to functions that take as arguments the things they need to do their behavior.

File: shiny-api/api/handlers.go [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+	}
+}
+
+type ThingInput struct {
+	Name    string  `json:"name"`
+	Foo     float64 `json:"foo"`
+	Version string  `json:"version"`
+}
+
+func CodeOrDefault(err error, def int) int {
```
Comment (varun06) [inline shiny-api/api/handlers.go]: Are the all exported?? 
if yes, then go lint/vet will complain about comments for these functions..

File: shiny-api/api/handlers.go [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+	type coder interface {
```
Comment (varun06) [inline shiny-api/api/handlers.go]: any particular reason to define this interface here?

**Pull request merged**